vignettes/cookbook.Rmd

```diff
@@ -0,0 +1,221 @@
+  returning = porcelain::porcelain_returning_json("lmCoef", schema_root))
+```
+
+The input schema, `lmInputs.json` is
+
+```{r, echo = FALSE, results = "asis"}
+show_file(file.path(schema_root, "lmInputs.json"), "json")
+```
+
+while the response schema `lmCoef.json` is
+
+```{r, echo = FALSE, results = "asis"}
+show_file(file.path(schema_root, "lmCoef.json"), "json")
+```
+
+These are both fairly strict schemas using both `required` and `additionalProperties`. You might want to be more permissive, but we find that strictness here pays off later.
+
+```{r}
+api <- porcelain::porcelain$new(validate = TRUE)$handle(endpoint_lm)
+```
+
+To exercise the API endpoint we need to pass in our input JSON (not an R object).
+
+```{r}
+json <- jsonlite::toJSON(data)
+json
+```
+
+```{r}
+api$request("POST", "/lm", body = json, content_type = "application/json")
+```
+
+# POST endpoint, inputs as binary, returning binary
+
+(This example also shows off a few other features)
+
+Handling binary inputs and outputs is supported, provided that you can deal with them in R.  In this example we'll use R's serialisation format (rds; see `?serialize` and `?saveRDS`) as an example, but this approach would equally work with excel spreadsheets, zip files or any other non-text data that you work with.
+
+In this example we'll take some serialised R data and create a png plot as output.  We'll start by writing our target function:
+
+```{r}
+binary_plot <- function(data, width = 400, height = 400) {
+  data <- unserialize(data)
+  tmp <- tempfile(fileext = ".png")
+  on.exit(unlink(tmp))
+  png(tmp, width = width, height = height)
+  tryCatch(
+    plot(data),
+    finally = dev.off())
+  readBin(tmp, raw(), n = file.size(tmp))
+}
+```
+
+Here, we use `unserialize` to convert the incoming binary data into something usable, plot to a temporary file (which we clean up later, using `on.exit`). Using `tryCatch(..., finally = dev.off())` ensures that even if the plotting fails, the device will be closed.  Finally, `readBin` reads that temporary file in a raw vector.
+
+So, for example (using `str` to limit what is printed to screen)
+
+```{r}
+bin <- serialize(data, NULL)
+str(binary_plot(bin), vec.len = 10)
+```
+
+It's hard to tell this is a png, but the first few bytes give it away (the [magic number](https://en.wikipedia.org/wiki/List_of_file_signatures) `89 50 4e 47 0d 0a 1a 0a` is used at the start of all png files).
+
+```{r}
+endpoint_plot <- porcelain::porcelain_endpoint$new(
+  "POST", "/plot", binary_plot,
+  porcelain::porcelain_input_body_binary("data"),
+  returning = porcelain::porcelain_returning_binary())
+api <- porcelain::porcelain$new(validate = TRUE)$handle(endpoint_plot)
+```
+
+Making the request (again using `str` to prevent printing thousands of hex characters)
+
+```{r}
+str(api$request("POST", "/plot", body = bin,
+                 content_type = "application/octet-stream"),
+    vec.len = 10)
+```
```

Comment:
We could add a section on HEAD endpoints too as these aren't straightforward to implement

Reply:
Example of error case too?

Reply:
So many to add, these are all needed - error handling is going to be a big section I think

Reply:
https://mrc-ide.myjetbrains.com/youtrack/issue/RESIDE-260